### ISSUE TITLE
How to change metric in early stopping?

### ISSUE THREAD

**Issue by hdnh2006** [ultralytics/yolov5#5561]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi Glenn and ultralytics guys!

I am getting the following results while training a model from `yolov5s.pt` architecture:

**MAP 0.5:**
![image](https://user-images.githubusercontent.com/17271049/140731499-2861ec1d-48a2-4a48-89e0-c8b917c78471.png)

**Precision:**
![image](https://user-images.githubusercontent.com/17271049/140731887-d93b1ddf-8992-4718-bc5a-c184f78b7a04.png)

**Recall:**
![image](https://user-images.githubusercontent.com/17271049/140731945-9c2b4c4a-db0b-4e37-a48c-39b83883fd3b.png)

It is clearly a case of overfitting from the epoch 37 approx and I set a patience of 50 epochs. However, the process doesn't stop because (I think) it is reducing the loss while training
![image](https://user-images.githubusercontent.com/17271049/140732281-8957bb90-e152-4113-a251-efe4b0e6f554.png)

So, my question is, how do I change the target metric for early stopping? In my case I would like to set the precision or recall in validation set for example. That is:

`if map < map_early:
   stop
else:
   continue training...
`

**Comment by glenn-jocher** [comment]
@hdnh2006 early stopping works correctly. In your example 50 epochs have not elapsed since the best mAP.

**Comment by glenn-jocher** [comment]
@hdnh2006 you can reduce early stopping patience with the --patience argument, i.e. for 10 epochs:
```bash
python train.py --patience 10
```

**Comment by hdnh2006** [comment]
Thanks @glenn-jocher, It is running the epoch number 60 and from the epoch ~37 there's no improvement. It should stop from epoch 51 I guess.

Anyway, is there some way to change the target metric?

**Comment by glenn-jocher** [comment]
@hdnh2006 you can read some documentation on early stopping behavior here:
https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping

You're not understanding patience. If patience is 50 and best epoch was observed at epoch 100, training will continue for 50 epochs with no improvement before stopping at epoch 150.

**Comment by hdnh2006** [comment]
Got it! thank you.

**Comment by DoSquared** [comment]
How can I set patience such that the algorithm does not stop till all epochs are finished? is it 0?

**Comment by glenn-jocher** [comment]
@DaliaMahdy you can set a higher number of epochs for early stopping, i.e. --patience 500 (default is 100), or you could disable it completely with --patience 0
```
python train.py --patience 0
```

**Comment by mnoroozii** [comment]
@glenn-jocher I want to change the metric for the early stop as well. for example, using recall instead of map. Any suggestion?

**Comment by glenn-jocher** [comment]
@mnoroozii Currently, the early stopping in YOLOv5 is designed to monitor validation loss, not other metrics like recall or precision. Modifying the early stopping criterion to monitor a different metric such as recall would require custom changes to the training script. 

This involves adjusting the code where the early stopping condition is checked, to instead monitor your metric of choice. However, this feature is not supported out of the box and would require a good understanding of the codebase to implement correctly.

For the most accurate guidance on implementing such a feature, I recommend reviewing the training loop in the source code and considering how the validation metrics are tracked and evaluated. Remember, any modifications you make for your use case would be specific to your project and not supported by the official repository.

**Comment by 7rkMnpl** [comment]
can any one help me for yolov5 model how to link early stop and callback( custom callback) in . yolov5 from ultralytics

**Comment by glenn-jocher** [comment]
Hi @7rkMnpl,

To integrate a custom callback with early stopping in YOLOv5, you would need to modify the training script to include your custom callback logic. Here's a general outline of how you can achieve this:

1. **Create Your Custom Callback:**
   Define your custom callback class. For example, you might want to create a callback that monitors a specific metric and stops training based on that metric.

   ```python
   class CustomEarlyStopping:
       def __init__(self, patience=10, min_delta=0):
           self.patience = patience
           self.min_delta = min_delta
           self.best_score = None
           self.counter = 0

       def __call__(self, current_score):
           if self.best_score is None:
               self.best_score = current_score
           elif current_score < self.best_score + self.min_delta:
               self.counter += 1
               if self.counter >= self.patience:
                   return True
           else:
               self.best_score = current_score
               self.counter = 0
           return False
   ```

2. **Integrate the Callback into the Training Loop:**
   Modify the training loop in `train.py` to include your custom callback. You will need to check the callback condition at the end of each epoch.

   ```python
   from train import train

   # Initialize your custom callback
   custom_early_stopping = CustomEarlyStopping(patience=10, min_delta=0.01)

   # Modify the training loop to include the callback check
   for epoch in range(epochs):
       # Training code...
       
       # Calculate your custom metric (e.g., recall)
       current_score = calculate_recall()

       # Check the custom early stopping condition
       if custom_early_stopping(current_score):
           print(f"Early stopping at epoch {epoch}")
           break
   ```

3. **Run Your Training Script:**
   Execute your modified training script to train your YOLOv5 model with the custom early stopping callback.

This is a basic example to get you started. Depending on your specific requirements, you might need to adjust the callback logic and how you integrate it into the training loop.

Feel free to ask if you have any further questions or need additional assistance. Happy training! 😊

**Comment by 7rkMnpl** [comment]
> Hi @7rkMnpl,
> 
> To integrate a custom callback with early stopping in YOLOv5, you would need to modify the training script to include your custom callback logic. Here's a general outline of how you can achieve this:
> 
> 1. **Create Your Custom Callback:**
>    Define your custom callback class. For example, you might want to create a callback that monitors a specific metric and stops training based on that metric.
>    ```python
>    class CustomEarlyStopping:
>        def __init__(self, patience=10, min_delta=0):
>            self.patience = patience
>            self.min_delta = min_delta
>            self.best_score = None
>            self.counter = 0
>    
>        def __call__(self, current_score):
>            if self.best_score is None:
>                self.best_score = current_score
>            elif current_score < self.best_score + self.min_delta:
>                self.counter += 1
>                if self.counter >= self.patience:
>                    return True
>            else:
>                self.best_score = current_score
>                self.counter = 0
>            return False
>    ```
> 2. **Integrate the Callback into the Training Loop:**
>    Modify the training loop in `train.py` to include your custom callback. You will need to check the callback condition at the end of each epoch.
>    ```python
>    from train import train
>    
>    # Initialize your custom callback
>    custom_early_stopping = CustomEarlyStopping(patience=10, min_delta=0.01)
>    
>    # Modify the training loop to include the callback check
>    for epoch in range(epochs):
>        # Training code...
>        
>        # Calculate your custom metric (e.g., recall)
>        current_score = calculate_recall()
>    
>        # Check the custom early stopping condition
>        if custom_early_stopping(current_score):
>            print(f"Early stopping at epoch {epoch}")
>            break
>    ```
> 3. **Run Your Training Script:**
>    Execute your modified training script to train your YOLOv5 model with the custom early stopping callback.
> 
> This is a basic example to get you started. Depending on your specific requirements, you might need to adjust the callback logic and how you integrate it into the training loop.
> 
> Feel free to ask if you have any further questions or need additional assistance. Happy training! 😊

thank you..

**Comment by 7rkMnpl** [comment]
import os
import torch
import yaml
from yolov5 import train
from yolov5 import train as yolov5_train
from yolov5.utils.plots import plot_results
from yolov5.utils.callbacks import Callbacks

class EarlyStopping:
    def __init__(self, patience=5, verbose=False, delta=0):
        self.patience = patience
        self.verbose = verbose
        self.delta = delta
        self.best_score = None
        self.early_stop = False
        self.counter = 0

    def __call__(self, val_loss, model):
        score = -val_loss
        if self.best_score is None:
            self.best_score = score
            self.save_checkpoint(val_loss, model)
        elif score < self.best_score + self.delta:
            self.counter += 1
            if self.verbose:
                print(f'EarlyStopping counter: {self.counter} out of {self.patience}')
            if self.counter >= self.patience:
                self.early_stop = True
        else:
            self.best_score = score
            self.save_checkpoint(val_loss, model)
            self.counter = 0

    def save_checkpoint(self, val_loss, model):
        if self.verbose:
            print(f'Validation loss decreased ({self.best_score:.6f} --> {val_loss:.6f}).  Saving model ...')
        torch.save(model.state_dict(), 'checkpoint.pt')
       

# Custom training function
def custom_train(data, epochs, batch_size, img_size, patience):
    # Load data configuration
    with open(data) as f:
        data_dict = yaml.safe_load(f)
    
    # Initialize model
    model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True)
    
    # Set up early stopping
    early_stopping = EarlyStopping(patience=patience, verbose=True)
    
    # Train model
    for epoch in range(epochs):
        # Run training
        train.run(
            img_size=img_size,
            batch_size=batch_size,
            epochs=1,  # Run one epoch at a time
            data=data,
            weights='yolov5s.pt',
            save_period=1,
            project='yolov5/runs/train',
            name=f'custom_train_{epoch}'
        )
        
        # Load validation results
        results_file = Path('/home/yolov5/runs/train/custom_train_{epoch}/results.txt')
        if results_file.exists():
            with open(results_file) as f:
                results = f.readlines()
                val_loss = float(results[-1].split()[3])  # Extracting val_loss from results file
                
            # Check early stopping
            early_stopping(val_loss, model, epoch)
            if early_stopping.early_stop:
                print("Early stopping")
                break
    
    # Load the last checkpoint with the best model
    best_epoch = early_stopping.best_score
    best_checkpoint = f'checkpoint_epoch_{best_epoch}.pt'
    model.load_state_dict(torch.load(best_checkpoint))
    return model

# Paths
data_path = '/home/yolov5/dataset/data.yaml'

# Training parameters
epochs = 75
batch_size = 8
img_size = 512
patience = 5

# Train the model with early stopping
model = custom_train(data=data_path, epochs=epochs, batch_size=batch_size, img_size=img_size, patience=patience)

# Save the final model
model.save('best_model.pt')

**Comment by 7rkMnpl** [comment]
this is my code how to modify this for above input

**Comment by 7rkMnpl** [comment]
here is my code
import os
import torch
import yaml
from yolov5 import train
from yolov5 import train as yolov5_train
from yolov5.utils.plots import plot_results
from yolov5.utils.callbacks import Callbacks

class EarlyStopping:
    def __init__(self, patience=5, verbose=False, delta=0):
        self.patience = patience
        self.verbose = verbose
        self.delta = delta
        self.best_score = None
        self.early_stop = False
        self.counter = 0

    def __call__(self, val_loss, model):
        score = -val_loss
        if self.best_score is None:
            self.best_score = score
            self.save_checkpoint(val_loss, model)
        elif score < self.best_score + self.delta:
            self.counter += 1
            if self.verbose:
                print(f'EarlyStopping counter: {self.counter} out of {self.patience}')
            if self.counter >= self.patience:
                self.early_stop = True
        else:
            self.best_score = score
            self.save_checkpoint(val_loss, model)
            self.counter = 0

    def save_checkpoint(self, val_loss, model):
        if self.verbose:
            print(f'Validation loss decreased ({self.best_score:.6f} --> {val_loss:.6f}).  Saving model ...')
        torch.save(model.state_dict(), 'checkpoint.pt')
       

# Custom training function
def custom_train(data, epochs, batch_size, img_size, patience):
    # Load data configuration
    with open(data) as f:
        data_dict = yaml.safe_load(f)
    
    # Initialize model
    model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True)
    
    # Set up early stopping
    early_stopping = EarlyStopping(patience=patience, verbose=True)
    
    # Train model
    for epoch in range(epochs):
        # Run training
        train.run(
            img_size=img_size,
            batch_size=batch_size,
            epochs=1,  # Run one epoch at a time
            data=data,
            weights='yolov5s.pt',
            save_period=1,
            project='yolov5/runs/train',
            name=f'custom_train_{epoch}'
        )
        
        # Load validation results
        results_file = Path('/home/yolov5/runs/train/custom_train_{epoch}/results.txt')
        if results_file.exists():
            with open(results_file) as f:
                results = f.readlines()
                val_loss = float(results[-1].split()[3])  # Extracting val_loss from results file
                
            # Check early stopping
            early_stopping(val_loss, model, epoch)
            if early_stopping.early_stop:
                print("Early stopping")
                break
    
    # Load the last checkpoint with the best model
    best_epoch = early_stopping.best_score
    best_checkpoint = f'checkpoint_epoch_{best_epoch}.pt'
    model.load_state_dict(torch.load(best_checkpoint))
    return model

# Paths
data_path = '/home//yolov5/dataset/data.yaml'

# Training parameters
epochs = 75
batch_size = 8
img_size = 512
patience = 5

# Train the model with early stopping
model = custom_train(data=data_path, epochs=epochs, batch_size=batch_size, img_size=img_size, patience=patience)

# Save the final model
model.save('best_model.pt')

**Comment by glenn-jocher** [comment]
Hi @7rkMnpl,

Thank you for sharing your code! To integrate a custom callback for early stopping based on a different metric (e.g., recall) in your training loop, you can modify your `EarlyStopping` class and the training loop accordingly. Here's how you can do it:

1. **Modify the `EarlyStopping` Class:**
   Update the class to accept a custom metric (e.g., recall) instead of validation loss.

   ```python
   class EarlyStopping:
       def __init__(self, patience=5, verbose=False, delta=0):
           self.patience = patience
           self.verbose = verbose
           self.delta = delta
           self.best_score = None
           self.early_stop = False
           self.counter = 0

       def __call__(self, metric, model):
           score = metric
           if self.best_score is None:
               self.best_score = score
               self.save_checkpoint(metric, model)
           elif score < self.best_score + self.delta:
               self.counter += 1
               if self.verbose:
                   print(f'EarlyStopping counter: {self.counter} out of {self.patience}')
               if self.counter >= self.patience:
                   self.early_stop = True
           else:
               self.best_score = score
               self.save_checkpoint(metric, model)
               self.counter = 0

       def save_checkpoint(self, metric, model):
           if self.verbose:
               print(f'Metric improved ({self.best_score:.6f} --> {metric:.6f}). Saving model ...')
           torch.save(model.state_dict(), 'checkpoint.pt')
   ```

2. **Update the Training Loop:**
   Modify the training loop to calculate and use the custom metric (e.g., recall) for early stopping.

   ```python
   from pathlib import Path
   import torch
   import yaml
   from yolov5 import train
   from yolov5.utils.plots import plot_results

   # Custom training function
   def custom_train(data, epochs, batch_size, img_size, patience):
       # Load data configuration
       with open(data) as f:
           data_dict = yaml.safe_load(f)
       
       # Initialize model
       model = torch.hub.load('ultralytics/yolov5', 'yolov5s', pretrained=True)
       
       # Set up early stopping
       early_stopping = EarlyStopping(patience=patience, verbose=True)
       
       # Train model
       for epoch in range(epochs):
           # Run training
           train.run(
               img_size=img_size,
               batch_size=batch_size,
               epochs=1,  # Run one epoch at a time
               data=data,
               weights='yolov5s.pt',
               save_period=1,
               project='yolov5/runs/train',
               name=f'custom_train_{epoch}'
           )
           
           # Load validation results
           results_file = Path(f'/home/yolov5/runs/train/custom_train_{epoch}/results.txt')
           if results_file.exists():
               with open(results_file) as f:
                   results = f.readlines()
                   recall = float(results[-1].split()[4])  # Extracting recall from results file
                   
               # Check early stopping
               early_stopping(recall, model)
               if early_stopping.early_stop:
                   print("Early stopping")
                   break
       
       # Load the last checkpoint with the best model
       model.load_state_dict(torch.load('checkpoint.pt'))
       return model

   # Paths
   data_path = '/home/yolov5/dataset/data.yaml'

   # Training parameters
   epochs = 75
   batch_size = 8
   img_size = 512
   patience = 5

   # Train the model with early stopping
   model = custom_train(data=data_path, epochs=epochs, batch_size=batch_size, img_size=img_size, patience=patience)

   # Save the final model
   model.save('best_model.pt')
   ```

This code will now use recall as the metric for early stopping. Adjust the index in `results[-1].split()[4]` to match the position of the recall value in your results file.

Feel free to reach out if you have any further questions or need additional assistance. Happy training! 😊

**Comment by 7rkMnpl** [comment]
thank you..

**Comment by glenn-jocher** [comment]
You're welcome, @7rkMnpl!

If you have any further questions or run into any issues, feel free to ask. We're here to help! 😊

Just a quick reminder: if you encounter any bugs or issues, please ensure you are using the latest version of YOLOv5 and all related packages. This helps us ensure that any problems are not due to outdated software.

Happy training, and best of luck with your project! If you have any more questions, the YOLO community and the Ultralytics team are always here to support you.

**Comment by 7rkMnpl** [comment]
how does the yolov11 works for early stop for custom train code? should we write early stop code # patience does not helps, custom early stop did not help me.. any other alternatives for early stop code?

**Comment by pderrenger** [comment]
For YOLOv5, early stopping is based on validation loss. If custom methods aren't working, ensure you're using the latest version. You might need to adjust your patience settings or modify the training loop to suit your needs. If issues persist, please provide more details.
